### PR TITLE
🔒 feature(auth): LDAP Authentication

### LINKED ISSUE #1530: Enhancement:  Support for authentication via LDAP
### What features would you like to see added?

In addition to the authentication methods already provided, it would be nice to have the option to authenticate via LDAP (i.e. in combination with an OpenLDAP or AD instance)

### More details

This feature would be useful i. e. for institutions like universities which already provide LDAP authentication for other tools and platforms for their students/customers.

### Which components are impacted by your request?

_No response_

### Pictures

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by ohneda**
## Summary

This enhancement introduces LDAP authentication for login using email and password, ensuring that companies using LDAP servers can integrate their existing authentication infrastructure seamlessly with LibreChat.

Given that many companies, including mine 😁, still use LDAP servers, I think this feature is in demand.

Closes #1530 

## Changes

- Installed [passport-ldap](https://www.passportjs.org/packages/passport-ldapauth/)
- Added `LDAPStrategy` and `requireLdapAuth`
    * This feature is not for adding Social login, but for adding Email and Password authentication mechanisms. Therefore I did not add the LDAPStrategy to `api/server/socilaLogins.js`.
- Modified the login authentication (/login) to use LDAP if LDAP settings are provided in the `.env` file
- When LDAP authentication is used, `registrationEnabled` is always set to `false`.

## Change Type

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Testing

 I have manually verified the following:

- LDAP authentication is used when `LDAP_URL`, `LDAP_BIND_DN`, and `LDAP_USER_SEARCH_BASE` are specified.
- Local authentication is used when the above LDAP settings are not specified.
- When `LDAP_SEARCH_FILTER` is not specified, the system defaults to using the LDAP `mail` attribute to search for users.
- When a user logs in with LDAP authentication for the first time, a user is created with `provider=ldap` and `ldapId = ${uid}` of the LDAP server.

## Checklist

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have made pertinent documentation changes
- [x] My changes do not introduce new warnings
- [x] Local unit tests pass with my changes

### CODE REVIEW

**Comment by ohneda** [comment]
@danny-avila 

FYI: 

I saw [this proposal](https://github.com/danny-avila/LibreChat/issues/1530) associated with this PR

I've tested it using my company's LDAP server, but if you don't have an LDAP environment, you can test it using [slapd-server-mock](https://github.com/docker-ThoTeam/slapd-server-mock). This mock LDAP server contains test data by default.

It's basically as written in the README, but since I'm using an M1 Mac, I had to specify the `platform` in an arg like this when starting it up:

```
docker pull thoteam/slapd-server-mock --platform linux/amd64
docker run --rm -d -p 389:389 -p 636:636 --name my_slapd_mock --platform linux/amd64 thoteam/slapd-server-mock
```
If you want to test this PR using [the test data](https://github.com/docker-ThoTeam/slapd-server-mock/blob/main/bootstrap/data.ldif.TEMPLATE) included in the container by default, add the following to your .env file:

```env
LDAP_URL=ldap://localhost:389
LDAP_BIND_DN=cn=admin,dc=ldapmock,dc=local
LDAP_BIND_CREDENTIALS=adminpass
LDAP_USER_SEARCH_BASE=dc=ldapmock,dc=local
LDAP_SEARCH_FILTER=mail={{username}}
LDAP_CA_CERT_PATH=
```
Then you can log in with a user included in [the mock data](https://github.com/docker-ThoTeam/slapd-server-mock/blob/main/bootstrap/data.ldif.TEMPLATE). The password for all users is `password`. For example, if you log in as [developer.one@ldapmock.local](mailto:developer.one@ldapmock.local), you will log in as a user named "developer One" and the user's data will be created in the mongodb

Hope this helps

File: api/server/index.js [hunk 1]
```diff
@@ -15,7 +15,7 @@ const AppService = require('./services/AppService');
 const noIndex = require('./middleware/noIndex');
 const { isEnabled } = require('~/server/utils');
 const { logger } = require('~/config');
-
+const { ldapLogin } = require('~/strategies');
```
Comment (danny-avila) [inline api/server/index.js:18]: please note import order:

- npm packages first, 
     - from shortest line (top) to longest (bottom)

- followed by typescript types (mainly data-provider and client workspaces)
     - longest line (top) to shortest (bottom)
     - types from package come first

- lastly, local imports
     - longest line (top) to shortest (bottom)

**Review by danny-avila — APPROVED** [review]

**Comment by danny-avila** [comment]
Off topic but this is sad :( 

https://github.com/ldapjs/node-ldapjs/blob/8ffd0bc9c149088a10ec4c1ec6a18450f76ad05d/README.md

File: api/server/index.js (hunk 1 above)
Comment (ohneda) [reply api/server/index.js:18]: Thanks, I'll keep that in mind for next time!

Comment (danny-avila) [reply api/server/index.js:18]: Thank you! It helps manage all the boilerplate

**Comment by felixthekraut** [comment]
Hey y'all ended up here after trying to set AD up because I saw it in the config and was wondering why I couldn't get it to work. Brand spanking new feature - exciting!

I don't see it on the list for the 0.7.3 milestone - are you planning on adding this feature to that release or a later one?

**Pull request merged**